src/lib.rs

```diff
@@ -167,15 +167,21 @@ impl<T: ?Sized> TakeCell<T> {
         // This is also the only place (again, aside from `steal`) where we use/provide
         // a reference to the underlying value.
         //
-        // Since this `compare_exchange` only changes the value from `false` to `true`,
-        // it can only succeed once. This guarantees that the returned reference is
+        // Since this `swap` only changes the value from `false` to `true`, it can only
+        // return `false` once. This guarantees that the returned reference is
         // unique.
-        match self
-            .taken
-            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
-        {
-            Ok(_) => Some(unsafe { &mut *self.value.get() }),
-            Err(_) => None,
+        //
+        // `Relaxed` ordering is ok to use here, because when `TakeCell` is shared we
+        // only allow one (1) thread to access the protected memory, so there is no need
+        // to synchronize the memory between threads. When `TakeCell` is not shared and
+        // can be accessed with `get`, the thread that is holding `&mut TakeCell<_>`
+        // must have already synchronized itself with other threads so, again, there is
+        // no need for additional synchronization here.
+        match self.taken.swap(true, Ordering::Relaxed) {
+            // The cell was previously taken
+            true => None,
+            // The cell wasn't takes before, so we can take it
+            false => Some(unsafe { &mut *self.value.get() }),
```

Comment:
Need at least `fence(Acquire)` to prevent memory load and write instructions to happen before the check.

Reply:
Thanks, I'll fix this.

Is `fence(Acquire)` any better than using `swap` with `Acquire`?

Reply:
Honestly, here my knowledge is a bit flawed.

As I understand this:
`heal` would have `Release` semantics, but it uses mutable reference, so there **was** at least `Release` ordering that made the mutable reference safe to produce. That `Release` was not bound to this atomic value, so `fence` must be used.

Reply:
I'm confused now. I've tried to use [`loom`](https://github.com/tokio-rs/loom) (btw it's a pain to integrate) to test the current implementation and... it seems to work? 🤔 

Maybe I've accidentally synchronized the threads too much, but I can't yet figure out how to break the current implementation...

Reply:
So, I finally think that I understood why does this work.
> All modifications to any particular atomic variable occur in a total order that is specific to this one atomic variable.
> <https://en.cppreference.com/w/cpp/atomic/memory_order>

This means that only one `swap` can ever return `false`. While the cell can be reordered before the swap it's still fine since only one thread will get `false` and do anything with the cell.